src/util.c

```diff
@@ -344,6 +344,74 @@ float sanitize_scale(float scale) {
 	return max(0.1, scale);
 }
 
+bool point_in_ellipse(complex p, Ellipse e) {
+	double Xp = creal(p);
+	double Yp = cimag(p);
+	double Xe = creal(e.origin);
+	double Ye = cimag(e.origin);
+	double a = e.angle;
+
+	return (
+		pow(cos(a) * (Xp - Xe) + sin(a) * (Yp - Ye), 2) / pow(creal(e.axes)/2, 2) +
+		pow(sin(a) * (Xp - Xe) - cos(a) * (Yp - Ye), 2) / pow(cimag(e.axes)/2, 2)
+	) <= 1;
+}
+
+// Is the point of shortest distance between the line through a and b
+// and a point c between a and b and closer than r?
+// if yes, return f so that a+f*(b-a) is that point.
+// otherwise return -1.
+double lineseg_circle_intersect(LineSegment seg, Circle c) {
+	complex m, v;
+	double projection, lv, lm, distance;
+
+	m = seg.b - seg.a; // vector pointing along the line
+	v = seg.a - c.origin; // vector from circle to point A
+
+	lv = cabs(v);
+	lm = cabs(m);
+
+	if(lm == 0) {
```

Comment:
This thing prevents it to work for the case of zero segment width. the mystery check should go before this!